### PR TITLE
watchtower

### PR DESCRIPTION
**Pull request by eezcjkr**
resolves https://github.com/leapdao/bridge-ui/issues/108

### CODE REVIEW

**Comment by StevenJNPearce** [comment]
Fixes: #108 
@eezcjkr This is ready for review.

File: src/routes/status.tsx [hunk 1]
```diff
@@ -6,21 +6,231 @@
...
+        ).hash();
+        const exitTx = await explorer.getTransaction(exitTxHash);
+        if (exitTx === undefined) {
+          exit.suspect = true;
+        }
+        return exit;
+      })
+    );
+
+    const uTxos = await web3.plasma.instance.getUnspent('0x0'); //TODO: fix when leap web3 extension updated
```
Comment (StevenJNPearce) [inline src/routes/status.tsx]: This will need to be fixed, see: https://github.com/leapdao/leap-core/pull/67

**Comment by johannbarbie** [comment]
getting this one when doing `CONFIG=testnet yarn start`
```
ERROR in /Users/johba/dev/bridge-ui/src/routes/status.tsx
./src/routes/status.tsx
[tsl] ERROR in /Users/johba/dev/bridge-ui/src/routes/status.tsx(222,12)
      TS2322: Type '{ style: { paddingTop: string; }; bordered: true; dataSource: ColorDetails[]; columns: ColumnProps<any>[]; expandedRowRender: (record: any) => Element | "No open exits" | "No open exits for this token"; expandIcon: (props: any) => Element; title: () => Element; }' is not assignable to type 'IntrinsicAttributes & IntrinsicClassAttributes<Table<any>> & Pick<{ readonly children?: ReactNode; readonly prefixCls?: string; readonly dropdownPrefixCls?: string; readonly rowSelection?: TableRowSelection<...>; ... 32 more ...; readonly style?: CSSProperties; } & Pick<...>, "footer" | ... 24 more ... | "bodyStyle"...'.
  Property 'expandIcon' does not exist on type 'IntrinsicAttributes & IntrinsicClassAttributes<Table<any>> & Pick<{ readonly children?: ReactNode; readonly prefixCls?: string; readonly dropdownPrefixCls?: string; readonly rowSelection?: TableRowSelection<...>; ... 32 more ...; readonly style?: CSSProperties; } & Pick<...>, "footer" | ... 24 more ... | "bodyStyle"...'.
```

**Comment by StevenJNPearce** [comment]
@johannbarbie Did you run `yarn` to update the packages? I think that error was due to a bug in antd, solved by updating to 3.13.2

**Comment by johannbarbie** [comment]
> @johannbarbie Did you run `yarn` to update the packages? I think that error was due to a bug in antd, solved by updating to 3.13.2

didn't 😁 thx

**Comment by johannbarbie** [comment]
i've synced the latest node v3.3.4, but when opening status page the spinner keeps spinning:
screen shot 2019-02-20 at 11 45 08 am

**Comment by StevenJNPearce** [comment]
> i've synced the latest node v3.3.4, but when opening status page the spinner keeps spinning:
> screen shot 2019-02-20 at 11 45 08 am

No idea as i can't reproduce it; with metamask on Rinkeby using that same url for the node, the table loads for me.

**Comment by johannbarbie** [comment]
i'm also on rinkeby with my metamask 🤔 i have some completed exits with that account already. could that be the issue?

**Comment by johannbarbie** [comment]
i've checked out the repo one more time like this:
```
git checkout -b StevenJNPearce-watchtower master
git pull https://github.com/StevenJNPearce/bridge-ui.git watchtower
```
verified that all files are like in the pr.
i've also tried a different account that has not interacted with the chain yet. 
still spinning 🤷‍♂️

**Comment by johannbarbie** [comment]
i've logged `console.log(this.iContract);` in exitHandler.js `watchExits()`, and it is undefined

**Comment by StevenJNPearce** [comment]
Taking another look at it, it's not reliable; if I refresh the page 20% of the time it is getting stuck on spinning for me. Code is a bit convoluted so I'm going to try and simplify it to make it easier to figure out where things are going wrong.

**Comment by johannbarbie** [comment]
i think it is a timing issue with injection.
when page is loaded, then `this.iContract` is undefined, but when i check 1 sec later, then it is defined.

**Comment by sunify** [comment]
`iContract` is a computed property that depends on observable value (injected provider). You should use mobx's `reaction` or `autorun` to track when it has the value

Smth like

```es6
autorun(() => {
  if (this.iContract) {
    // do contract calls
  }
})
```

**Comment by sunify** [comment]
But why you're using `iContract`? It depends on injected provider and won't work without metamask.

File: src/stores/exitHandler.ts [hunk 2]
```diff
@@ -110,10 +132,20 @@ export default class ExitHandler extends ContractStore {
...
+    .on('data', async (event) => {
+      const utxoId = new Outpoint(event.returnValues.txHash, parseInt(event.returnValues.outIndex)).getUtxoId();
+      const exit = await this.iContract.methods.exits(this.web3.root.instance.utils.toHex(utxoId)).call();
+      this.Exits.push({ event, exit, suspect: false });
+    })
+  }
+
   @autobind
   @action
   private setAddress() {
```
Comment (sunify) [inline src/stores/exitHandler.ts]: You don't need this method. To have `address` property you can defined computed getter

```es6
@computed
get address() {
  return this.plasmaConfig.exitHandlerAddr;
}
```

and change reaction in constructor to
```
reaction(this.address, this.watchExits)
```

**Comment by StevenJNPearce** [comment]
> But why you're using `iContract`? It depends on injected provider and won't work without metamask.

Yea, i've seen that; it should just be `contract`

**Comment by StevenJNPearce** [comment]
Root web3 is not receiving events. i.e. change `iContract` to `contract` and `.on` never fires. It works ok if you do getPastEvents.

**Comment by StevenJNPearce** [comment]
I've updated it to use getPastEvents; it's not ideal but it works.

File: src/stores/exitHandler.ts [hunk 3]
```diff
@@ -18,9 +17,14 @@ import Transactions from '../components/txNotification/transactions';
 import { InflightTxReceipt } from '../utils/types';
 import Web3Store from './web3/';
 import PlasmaConfig from './plasmaConfig';
+import { Outpoint, Unspent } from 'leap-core';
+
 
 export default class ExitHandler extends ContractStore {
 
+  @observable
```
Comment (eezcjkr) [inline src/stores/exitHandler.ts]: This has to go, not being used and no definition of ExitData breaks build.

**Comment by sunify** [comment]
@StevenJNPearce why are you using force push here for each change? All commits will be squashed anyway, so you don't need to keep it in one commit or something like that.

Really hard to follow what's going on here

File: src/stores/exitHandler.ts (hunk 3 above)
Comment (StevenJNPearce) [reply src/stores/exitHandler.ts]: fixed.

**Comment by StevenJNPearce** [comment]
> @StevenJNPearce why are you using force push here for each change? All commits will be squashed anyway, so you don't need to keep it in one commit or something like that.
> 
> Really hard to follow what's going on here

Apologies; it's a habbit I picked up from being the preferred workflow on another project I contribute.

**Comment by johannbarbie** [comment]
i've tried to test locally again, but see this:
screen shot 2019-02-21 at 3 19 52 pm

**Comment by StevenJNPearce** [comment]
@johannbarbie This happens when for an unknown reason yarn installs web3@1.0.34 instead of web3@1.0.33 .Try delete node_modules and run `yarn` again.

**Comment by johannbarbie** [comment]
🤔 
```
a1708:bridge-ui johba$ cat node_modules/web3-core/package.json | grep version
    "version": "1.0.0-beta.33",
```

on first load:
- only spinner

on second load:
- error as above

**Comment by johannbarbie** [comment]
ok, got it. the problem was metamask.

File: src/routes/status.tsx [hunk 4]
```diff
@@ -6,21 +6,256 @@
...
+  }
+
+  @observable exitDetails: ExitData[] = [];
+  @observable colorDetails: ColorDetails[] = [];
+
+  @action
+  private getData = async (): Promise<void> => {
+    const { exitHandler, explorer, web3, tokens } = this.props;
+    const exits = []
+    const events = await exitHandler.contract.getPastEvents('ExitStarted', {fromBlock: 0})
```
Comment (johannbarbie) [inline src/routes/status.tsx]: this will probably not work on main-net. better to start from deployment of contract like this: https://github.com/leapdao/leap-node/blob/master/src/bridgeState.js#L75-L77
and then go in batches like here: https://github.com/leapdao/leap-node/blob/master/src/utils/ContractsEventsSubscription.js

**Comment by johannbarbie** [comment]
tested again, looks awesome already 👍 

found a few things:
*1.* stake is in ETH, not in LEAP
screen shot 2019-02-23 at 11 57 01 am
*2.* I don't understand the table yet. currently there are no pending exits for token 0, so i would expect exit sum to be 0. 
screen shot 2019-02-23 at 11 59 09 am
*3.* I assume the chain is invalid because some leap have been exited + doublespent. but i don't know how much. is this calculation correct: `(732.73334 + 31.1902) - 762.71354000000000001 = 1.21 LEAP`. so i can assume that 1.21 LEAP are missing. but shouldn't i see at least some exit as `Suspicious` then?.

**Comment by StevenJNPearce** [comment]
1) Ok.
2) Exit sum being wrong is a bug, it's including finalized exits in the calculation.
3) Nested table is only showing open exits. I will add a 2nd nested table showing bad finalized exits.

**Comment by johannbarbie** [comment]
awesome! i've also released leap-core v0.28.1 with the getUnspentAll() included.

**Comment by StevenJNPearce** [comment]
@johannbarbie I've addressed the above issues. Do you have any ideas how to make this a bit more user-friendly? Is there any information in the tables that isn't necessary? Some better terminology? There is quite a lot on display; could be difficult for users to make sense of it all especially if they are not super familiar with plasma.

**Comment by johannbarbie** [comment]
i absolutely agree, it is a lot on display, and yet this slim implementation of the watch tower doesn't even deliver full assurance to the user.

the point is to display the solvency of the chain. and if `sum of UTXOs + exits > balance of the contract on root chain`, then not every-one will be able to get their money back.

I think we will have to think more about user experience. for now i can image a few quick-fixes:
1. remove stake, as it is a global parameter, it will be same for every-one anyways
2. tell people that it is not enough to rely on this page, but that they should sync their own node to be sure. (instead of inflating the supply, the plasma operator could as well do invalid spends, which is only verifiable by running a local node.)
3. ask them to sign up for the newsletter, so that we can warn them. 

I found another functional issue:
Instead of PriorityTimestamp (age of utxo) we should read the `priority` [value](https://github.com/leapdao/leap-contracts/blob/master/contracts/ExitHandler.sol#L124), which will tell us how much time is left until the exit is finalized. But i'm not even sure it can currently be read from the contract. I will open a separate issue.

I think the only thing you can do for now is **1.** and maybe 3. For 2 we will need to have instructions for users how to download the node and run bridge-ui along-side it.

**Comment by johannbarbie** [comment]
@eezcjkr maybe you have ideas? i remember you have been critical about the user experience of this earlier.

**Comment by sunify** [comment]
I have some thought. Will show shortly

**Comment by sunify** [comment]
Note: I will talk about information representation only

1. Good job @StevenJNPearce 

2. Based on Johann's comment and what i see, data density of exits table way too low. Basically it's a three values — amount, address and date spread on free space evenly. And you don't need to compare them (table — list formatted for easy values comparison). That reasons why it's hard to read and why a table is not good format for that data. So I've simplified it and turned into list. Status of an exit encoded with color of amount, but can be extended with icon (or emoji, we love emoji).

3. Timestamp. What exactly it represents? Based on answer, we need to use proper format for it. Like `n time until blah blah` or `n time since blah blah`

4. Sorting? Currently it sorted by timestamp in straight order. Is it ok?

image

If it's ok, I'll push this stuff

P.S.

Most of `2.` can be said about colors table as well, but more time needed for it, so wait it in ep. 2 :-) But no need to wait for it to merge this PR.

File: src/routes/status.tsx [hunk 5]
```diff
@@ -5,22 +5,287 @@
  * found in the LICENSE file in the root directory of this source tree.
  */
 
+import { Badge, Button, Spin, Table } from 'antd';
+import { ColumnProps } from 'antd/lib/table';
+import BN from 'bn.js';
```
Comment (sunify) [inline src/routes/status.tsx]: `bn.js` is not listed in dependencies. Things can break after some web3 update

File: src/routes/status.tsx [hunk 6]
```diff
@@ -5,22 +5,287 @@
...
+    const colors = await Promise.all(
+      addresses.map(async addr => {
+        return web3.plasma.instance.getColor(addr);
+      })
+    );
+    const colorDetails = await Promise.all(
+      colors.map(async color => {
+        const utxoSum = uTxos.reduce((acc, utXo) => {
+          return utXo.output.color === color
+            ? web3.root.instance.utils.toBN(utXo.output.value).add(acc)
```
Comment (sunify) [inline src/routes/status.tsx]: `web3.root.instance.utils.toBN` — just use it directly from `bn.js` :-)

File: src/routes/status.tsx [hunk 7]
```diff
@@ -5,22 +5,287 @@
...
+            : acc;
+        }, web3.root.instance.utils.toBN(0));
+
+        const exitSum = exits.reduce((acc, e) => {
+          return parseInt(e.exit.color) === color && !e.exit.finalized
```
Comment (sunify) [inline src/routes/status.tsx]: Don't forget to add second argument to `parseInt` or use `Number(e.exit.color)` instead

File: src/routes/status.tsx [hunk 8]
```diff
@@ -5,22 +5,287 @@
...
+    runInAction(() => {
+      this.exitDetails = exits;
+      this.colorDetails = colorDetails;
+    });
+  };
+
+  private exitDetailsRenderer = record => {
+    if (!this.exitDetails.length) return 'No open exits';
+    const openExits = this.exitDetails.filter(
+      e => !e.exit.finalized && parseInt(e.exit.color) === record.color
```
Comment (sunify) [inline src/routes/status.tsx]: `parseInt`

File: src/routes/status.tsx [hunk 9]
```diff
@@ -5,22 +5,287 @@
...
+    );
+    console.log('openexits', openExits);
+    const badFinalizedExits = this.exitDetails.filter(
+      e =>
+        parseFloat(e.exit.color) === record.color &&
```
Comment (sunify) [inline src/routes/status.tsx]: Why `parseFloat` here?

File: src/routes/status.tsx [hunk 10]
```diff
@@ -5,22 +5,287 @@
...
+    4: string;
+    5: string;
+  };
+  suspect: boolean;
+}
 
+@inject('exitHandler', 'explorer', 'web3', 'tokens', 'bridge')
 @observer
-export default class Status extends React.Component {
+export default class Status extends React.Component<StatusProps, {}> {
```
Comment (sunify) [inline src/routes/status.tsx]: Will be better to extract watchtower stuff from Status component (like Monitor)

File: src/routes/status.tsx [hunk 11]
```diff
@@ -5,22 +5,287 @@
...
+        ).getUtxoId();
+        const exit = await exitHandler.contract.methods
+          .exits(web3.root.instance.utils.toHex(utxoId))
+          .call();
+        let suspect;
+        const exitTxHash = Tx.exit(
+          new Input(
+            new Outpoint(
+              event.returnValues.txHash,
+              parseInt(event.returnValues.outIndex)
```
Comment (sunify) [inline src/routes/status.tsx]: `parseInt`

File: src/routes/status.tsx [hunk 12]
```diff
@@ -5,22 +5,287 @@
...
+      await new Promise(resolve => reaction(() => bridge.address, resolve));
+    const fromBlock = await bridge.contract.methods.genesisBlockNumber().call();
+    const events = await exitHandler.contract.getPastEvents('ExitStarted', {
+      fromBlock,
+    });
+    await Promise.all(
+      events.map(async event => {
+        const utxoId = new Outpoint(
+          event.returnValues.txHash,
+          parseInt(event.returnValues.outIndex)
```
Comment (sunify) [inline src/routes/status.tsx]: `parseInt`

File: src/routes/status.tsx [hunk 13]
```diff
@@ -5,22 +5,287 @@
...
+
+  @observable exitDetails: ExitData[] = [];
+  @observable colorDetails: ColorDetails[] = [];
+
+  @action
+  private getData = async (): Promise<void> => {
+    const { exitHandler, explorer, web3, tokens, bridge } = this.props;
+    const exits = [];
+    if (!bridge.address)
+      await new Promise(resolve => reaction(() => bridge.address, resolve));
```
Comment (sunify) [inline src/routes/status.tsx]: Seems strange. If you need exitHandler and bridge in this method, will be better to use `autorun`.

**getData**
```ts
private getData = async (): Promise<void> => {
  if (!bridge.address || !exitHandler.address) {
    return;
  }
  ... do fetching here
}
```

**constructor**
```ts
constructor(props) {
  super(props);
  autorun(this.getData);
}
```

**Comment by StevenJNPearce** [comment]
@sunify Awesome, some great points and a nice improvement. I've pushed this to feature/watchtower, so you can push your changes there, and then I will address the code comments so you won't need to deal with any merge conflicts.

**Comment by StevenJNPearce** [comment]
I continued working on the UX; I think it is much better now.

capture
